### DISCUSSION TITLE
INLA (cgeneric) implementation for alpha=2 now available

### DISCUSSION THREAD

**Discussion by AlexandreBSimas** [davidbolin/MetricGraph#7] [General]
We now have an implementation for exact Whittle-Matern fields with `alpha = 2` in `INLA` and `inlabru`. 

See [INLA vignette](https://davidbolin.github.io/MetricGraph/articles/inla_interface.html#an-example-with-alpha-2) or [inlabru vignette](https://davidbolin.github.io/MetricGraph/articles/inlabru_interface.html#an-example-with-alpha-2).

**Comment by AlexandreBSimas** [comment]
Observe that to run `alpha=2` now, you will need to install the `rSPDE` package from source. 

The details on how to do it are given here: https://davidbolin.github.io/rSPDE/articles/build_source.html

**Reply by karinalilleborge** [reply]
Dear Alexandre, 
I tried to install the rSPDE package as described in the link above, but received the following error message:


> Downloading GitHub repo davidbolin/rspde@devel-src
> ── R CMD build ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
> ✔  checking for file ‘/private/var/folders/dy/5rwl54sx6kv74t5tw02b6n6w0000gn/T/Rtmp4afTAV/remotes10193744cbf0f/davidbolin-rSPDE-8b3fa07/DESCRIPTION’ ...
> ─  preparing ‘rSPDE’:
> ✔  checking DESCRIPTION meta-information ...
> ─  cleaning src
> ─  checking for LF line-endings in source and make files and shell scripts
> ─  checking for empty or unneeded directories
> ─  looking to see if a ‘data/datalist’ file should be added
> ─  building ‘rSPDE_2.3.3.9000.tar.gz’
>    
> * installing *source* package ‘rSPDE’ ...
> ** using staged installation
> ** libs
> /usr/local/bin/g++ -Wl,-ld_classic -O2 -Wall -Wextra -fpic -g -I/usr/local/brewinclude -fopenmp -I/Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library/INLA/include/ -I/usr/local/include/eigen3/ -I/usr/include/eigen3/ -c cgeneric_mvnormdens.cpp -o cgeneric_mvnormdens.o
> make: /usr/local/bin/g++: No such file or directory
> make: *** [cgeneric_mvnormdens.o] Error 1
> ERROR: compilation failed for package ‘rSPDE’
> * removing ‘/Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library/rSPDE’
> * restoring previous ‘/Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library/rSPDE’
> Warning message:
> In i.p(...) :
>   installation of package ‘/var/folders/dy/5rwl54sx6kv74t5tw02b6n6w0000gn/T//Rtmp4afTAV/file101935a93ad6f/rSPDE_2.3.3.9000.tar.gz’ had non-zero exit status


Is it something wrong with my attempt, or is it the new version? I tried to do the 
`remotes::install_github("davidbolin/rspde", ref = "devel-src")`. 
Since it mentions the cgeneric_mvnormdens.cpp, I thought it might be related to the new implementation?

Thank you for implementing alpha=2. Looking forward to test it out:-)

Kind regards,
Karina Lilleborge

**Reply by AlexandreBSimas** [reply]
Hi Karina,

Really good that you are going to use it!!

From what I see it is not finding g++ on `/usr/local/bin` folder:
```
/usr/local/bin/g++: No such file or directory
```

I updated the instructions here `https://davidbolin.github.io/rSPDE/articles/build_source.html` recently, so maybe when you saw, it was not updated.

Basically, you need to make sure you have gcc-14 installed (for instance, using brew):
```
brew install gcc@14
```

Then, you need to link it to that folder `/usr/local/bin` to be able use our Makefile:

```
sudo ln -s $(brew --prefix gcc@14)/bin/gcc-14 /usr/local/bin/gcc
sudo ln -s $(brew --prefix gcc@14)/bin/g++-14 /usr/local/bin/g++
```

This only creates a symbolic link to that folder.

You will also need `eigen`:

```
brew install eigen
```

and the following symbolic links:


```
sudo ln -s $(brew --prefix eigen)/include/eigen3 /usr/local/include/eigen3
sudo ln -s $(brew --prefix)/lib /usr/local/brewlib
sudo ln -s $(brew --prefix)/include /usr/local/brewinclude
```

With all of this, you should be able to install.

Please, let me know if you encounter any problems.

**Reply by karinalilleborge** [reply]
Thank you for clarifying, Alexandre! I was missing the dependencies, indeed. The error when trying to install `rSPDE` with `remotes::install_github` is now related to Eigen instead of g++. 
I am trying to follow the instructions, but when I try to install it with brew, I get the following 

> Warning: eigen 3.4.0_1 is already installed and up-to-date.

Then when I try to link it with 
`sudo ln -s $(brew --prefix eigen)/include/eigen3 /usr/local/include/eigen3` 
it cannot find the file/directory:

> ln: /usr/local/include/eigen3: No such file or directory

I also tried to do the `brew reinstall eigen` as it suggested that with the warning, but nothing changed. 

I checked the location of `eigen` with `brew --prefix eigen` and got 

> /opt/homebrew/opt/eigen

but I must admit I am not sure if I understand what I should change in the linking (or if I should change something in the linking at all?). 
Also, I don't understand the section about the Makefile. Where is this Makefile located? In rSPDE (that you can clone from git) or is it related to something else?

**Reply by AlexandreBSimas** [reply]
Hi Karina,

From the error: `ln: /usr/local/include/eigen3: No such file or directory`

you probably don't have an `include` folder inside `/usr/local`. So, you can try
```
sudo mkdir -p /usr/local/include
```

and then try to link again. After that, you can try to install again.

All you need to do is to create the link, because the Makefile, used to compile (it is inside the src folder from the `devel-src` branch). If you link like that, you probably will not need to touch the Makefile. 

There is only one possibility, which is related to the intrinsic models, that sometimes make it annoying to compile, if you face that, the easiest way is to adjust the Makefile indeed. But let's not worry about this now, and there is a very good chance it will directly work.

Please, let me know if you still run into some problem.

**Reply by karinalilleborge** [reply]
I managed to make the link and also successfully downloaded rSPDE. 🥳 Thank you so much for the quick response and help.

**Reply by AlexandreBSimas** [reply]
Great!!! My pleasure!!
